File: src/main/java/math/cat/Pair.java
package math.cat;

import static math.cat.Base.equal;

import java.util.Map;

/**
 * Abstract pair class.
 * Implementations should implement two methods, and may be lazy or regular.
 * But in any case this is an unmodifiable pair.
 * 
 * @author Vlad Patryshev
 * All source code is stored on <a href="http://code.google.com/p/categories/">http://code.google.com/p/categories/</a>
 * 
 * @param <X> first component type
 * @param <Y> second component type
 */
public abstract class Pair<X, Y> implements Map.Entry<X, Y> {

  /**
   * @return first element of the pair.
   */
  public abstract X x();

  /**
   * @return second element of the pair.
   */
  public abstract Y y();

  /**
   * @return first element of the pair.
   */
  public X getKey() {
    return x();
  }

  /**
   * @return second element of the pair.
   */
  public Y getValue() {
    return y();
  }

  /**
   * Setter method; not implemented.
   *
   * @inheritDoc
   */
  public Y setValue(Y value) {
    throw new UnsupportedOperationException();
  }

  @Override
  public boolean equals(Object o) {
    return o instanceof Pair && equal(((Pair<?, ?>) o).x(), x()) && equal(((Pair<?, ?>) o).y(), y());
  }

  /**
   * Hashcode for a pair. HAS TO use the same algorithm as Math.Entry implementations,
   * or else oops.
   *
   * @return the hashcode...
   */
  @Override
  public int hashCode() {
    return Base.hashCode(x()) ^ Base.hashCode(y());
  }

  @Override
  public String toString() {
    return "(" + x() + "," + y() + ")";
  }
}

File: src/main/scala/math/cat/Pair.scala
package math.cat


import math.cat.Functions._

object Pair {

  /**
   * Builds a function that builds pairs attaching x as the first element.
   *
   * @tparam X first component type
   * @tparam Y second component type
   * @param x the element to attach.
   * @return a function y -> (x, y)
   */
  def withLeft[X, Y](x: X): Injection[Y, (X,Y)] = new Injection[Y, (X,Y)] {
    def apply(y: Y): (X, Y) = (x, y)
  }

  /**
   * Builds a function that builds pairs attaching y as the second element.
   *
   * @tparam X first component type
   * @tparam Y second component type
   * @param y the element to attach.
   * @return a function x -> (x, y)
   */
  def withRight[X, Y](y: Y): Injection[X, (X,Y)] = new Injection[X, (X,Y)] {
    def apply(x: X): (X, Y) = (x, y)
  }
}